src/main/java/org/gridsuite/directory/server/DirectoryService.java

```diff
@@ -610,4 +619,149 @@ public String getDuplicateNameCandidate(UUID directoryUuid, String elementName,
+
+                    // Combine parent and descendants into a single list
+                    List<DirectoryElementEntity> result = new ArrayList<>();
+                    result.add(entity);
+                    result.addAll(descendants);
+
+                    return result.stream().map(e -> e.stashElement(false, null));
+                })
+                .toList();
+
+        var parent = getElement(parentUuid);
+        notificationService.emitDirectoryChanged(
+                parentUuid,
+                parent.getElementName(),
+                userId,
+                null,
+                parent.getAccessRights().isPrivate(),
+                parentUuid == null,
+                NotificationType.UPDATE_DIRECTORY
+        );
+        directoryElementRepository.saveAll(entities);
+        emitDirectoryChangedNotification(parentUuid, userId);
+    }
+
+    public void stashElements(List<UUID> elementsUuid, String userId) {
+        // we add the same stash date to all the elements that are deleted together
+        LocalDateTime stashDate = LocalDateTime.now();
+        List<DirectoryElementEntity> entities = directoryElementRepository.findAllByIdInAndStashed(elementsUuid, false);
+        List<DirectoryElementEntity> notUpdatableEntities = new ArrayList<>();
+        List<DirectoryElementEntity> updatableEntities = getUpdatableEntities(entities, notUpdatableEntities, userId, true);
+
+        directoryElementRepository.saveAll(updatableEntities.stream()
+                .flatMap(entity -> {
+                    List<DirectoryElementEntity> descendants = directoryElementRepository.findAllDescendants(entity.getId(), userId);
+                    // Combine parent and descendants into a single list
+                    List<DirectoryElementEntity> result = new ArrayList<>();
+                    result.add(entity);
+                    result.addAll(descendants);
+                    return result.stream().map(e -> {
+                        DirectoryElementEntity stashedElement = e.stashElement(true, stashDate);
+                        if (Objects.equals(e.getType(), STUDY)) {
+                            notificationService.emitDeletedStudy(entity.getId(), userId);
+                        }
+                        return stashedElement;
+                    });
+                })
+                .toList());
+
+        updatableEntities.forEach(entity -> {
+            UUID parentUuid = getParentUuid(entity.getId());
+            notificationService.emitDirectoryChanged(
+                    parentUuid == null ? entity.getId() : parentUuid,
+                    entity.getName(),
+                    userId,
+                    null,
+                    entity.getIsPrivate(),
+                    parentUuid == null,
+                    parentUuid == null ? NotificationType.DELETE_DIRECTORY : NotificationType.UPDATE_DIRECTORY
+            );
+        });
```

Comment:
A simple for loop would be more efficient

Reply:
why ? i think foreach is more clean and readable